sigtool/sigtool.c

```diff
@@ -1566,6 +1573,404 @@ static int listsigs(const struct optstruct *opts, int mode)
+
+    if ((ret = cl_engine_set_str(engine, CL_ENGINE_TMPDIR, tempdir))) {
+        logg(LOGG_ERROR, "cli_engine_set_str(CL_ENGINE_TMPDIR) failed: %s\n", cl_strerror(ret));
+        goto done;
+    }
+
+    bRet = true;
+done:
+    return bRet;
+}
+
+static int vbadump(const struct optstruct *opts)
+{
+    int ret = -1;
+
+    struct cl_scan_options options;
+    struct cl_engine *engine = NULL;
+    ScanFile *sf             = NULL;
+
+#ifndef _WIN32
+    struct rlimit rlim;
+#endif
+
+    const char *filename = NULL;
+    ScanFiles scanFiles;
+
+    memset(&scanFiles, 0, sizeof(ScanFiles));
+
+    /* Initalize scan options struct */
+    memset(&options, 0, sizeof(struct cl_scan_options));
+
+    if ((ret = cl_init(CL_INIT_DEFAULT))) {
+        logg(LOGG_ERROR, "Can't initialize libclamav: %s\n", cl_strerror(ret));
+        ret = 2;
+        goto done;
+    }
+
+    if (!(engine = cl_engine_new())) {
+        logg(LOGG_ERROR, "Can't initialize antivirus engine\n");
+        ret = 2;
+        goto done;
+    }
+
+    cl_engine_set_clcb_file_props(engine, filePropsCallback);
+    cl_engine_set_clcb_vba(engine, vba_callback);
+
+    if (!setTempDir(engine, opts)) {
+        ret = 2;
+        goto done;
+    }
+
+    if ((ret = cl_engine_compile(engine)) != 0) {
+        logg(LOGG_ERROR, "Database initialization error: %s\n", cl_strerror(ret));
+        ret = 2;
+        goto done;
+    }
+
+#ifndef _WIN32
+    if (getrlimit(RLIMIT_FSIZE, &rlim) == 0) {
+        if (rlim.rlim_cur < (rlim_t)cl_engine_get_num(engine, CL_ENGINE_MAX_FILESIZE, NULL))
+            logg(LOGG_WARNING, "System limit for file size is lower than engine->maxfilesize\n");
+        if (rlim.rlim_cur < (rlim_t)cl_engine_get_num(engine, CL_ENGINE_MAX_SCANSIZE, NULL))
+            logg(LOGG_WARNING, "System limit for file size is lower than engine->maxscansize\n");
+    } else {
+        logg(LOGG_WARNING, "Cannot obtain resource limits for file size\n");
+    }
+#endif
+
+    /*Enable options that are needed to parse ole2 and ooxml files for vba macros.  */
+    options.parse |= CL_SCAN_PARSE_ARCHIVE;
+    options.parse |= CL_SCAN_PARSE_XMLDOCS;
+
+    options.general |= CL_SCAN_GENERAL_COLLECT_METADATA;
+
+    /*Have the engine unpack everything so that we don't miss anything.  We'll clean it up later.*/
+    options.parse = 0xffffffff;
+
+    /*Need to explicitly set this to always keep temps, since we are scanning extracted ooxml/ole2 files
+     * from our scan directory to print all the vba content.  Remove it later if leave-temps was not
+     * specified */
+    cl_engine_set_num(engine, CL_ENGINE_KEEPTMP, 1);
+
+    filename = optget(opts, "vba")->strarg;
+    if (NULL == filename) {
+        filename = optget(opts, "vba-hex")->strarg;
+    }
+    scanfile(filename, engine, opts, &options, &scanFiles);
+
+    for (sf = scanFiles.lst; NULL != sf;) {
+        scanfile(sf->fileName, engine, opts, &options, &scanFiles);
+        sf = removeScanFile(&scanFiles, sf);
+    }
```

Comment:
cl_scandesc_callback is already a recursive process, so I would think that the process of adding files from the json to scan later would not be required.  Maybe we can message offline about it so I can understand this better